esp-hal/src/i2c.rs

```diff
@@ -1093,7 +1203,114 @@ mod asynch {
     }
 }
 
+mod private {
+    use super::Operation;
+
+    #[derive(PartialEq)]
+    // This enum is used to keep track of the last/next operation that was/will be
+    // performed in an embedded-hal(-async) I2C::transaction. It is used to
+    // determine whether a START condition should be issued at the start of the
+    // current operation and whether a read needs an ack or a nack for the final
+    // byte.
+    pub enum OpKind {
+        Write,
+        Read,
+    }
+
+    #[doc(hidden)]
+    pub trait I2cOperation<'a> {
+        fn is_write(&self) -> bool;
+
+        fn is_read(&self) -> bool;
+
+        fn write_buffer(&self) -> Option<&'a [u8]>;
+
+        fn read_buffer(&mut self) -> Option<&mut [u8]>;
+
+        fn is_empty(&self) -> bool;
+
+        fn kind(&self) -> OpKind;
+    }
+
+    impl<'a> I2cOperation<'a> for Operation<'a> {
+        fn is_write(&self) -> bool {
+            matches!(self, Operation::Write(_))
+        }
+
+        fn is_read(&self) -> bool {
+            matches!(self, Operation::Read(_))
+        }
+
+        fn write_buffer(&self) -> Option<&'a [u8]> {
```

Comment:
See above, I'm not too fond of these. They make sense in public code, but these aren't intended to be public as far as I can tell.

Reply:
me too - but see above